R/bmerge.R

```diff
@@ -89,7 +89,9 @@ bmerge = function(i, x, icols, xcols, roll, rollends, nomatch, mult, ops, verbos
           # common case of ad hoc user-typed integers missing L postfix joining to correct integer keys
           # we've always coerced to int and returned int, for convenience.
           if (verbose) cat("Coercing double column i.",names(i)[ic]," (which contains no fractions) to type integer to match type of x.",names(x)[xc],".\n",sep="")
-          val = as.integer(i[[ic]])
+          val = i[[ic]]
+          # according to ?as.integer, it's faster than `as.integer()` and will preserve the attributes
+          storage.mode(val) = "integer"
```

Comment:
I think it's faster in the case `val` has attributes but slower when not -- same benchmark with much larger vector:

```
N = 1e7
x1 = as.double(sample(1:2e4, N, TRUE))
x2 <- .Date(x1)

f1 <- function(x) {
  out <- as.integer(x)
  attributes(out) <- attributes(x)
  out
}

f2 <- function(x) {
  out <- as.integer(x)
  if (!is.null(attributes(x))) attributes(out) <- attributes(x)
  out
}

f3 <- function(x) {
  storage.mode(x) <- 'integer'
  x
}

bench::mark(
  f1(x1),
  f2(x1),
  f3(x1)
)

bench::mark(
  f1(x2),
  f2(x2),
  f3(x2)
)
```

```
# A tibble: 3 x 13
  expression    min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
  <bch:expr> <bch:> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
1 f1(x1)       16ms 18.2ms      48.4    38.1MB     4.84    20     2      413ms
2 f2(x1)     16.4ms 17.6ms      56.1    38.1MB    11.8     19     4      339ms
3 f3(x1)     32.5ms 34.5ms      28.6    38.1MB     4.77    12     2      419ms
# … with 4 more variables: result <list>, memory <list>, time <list>,
#   gc <list>
```

```
# A tibble: 3 x 13
  expression    min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
  <bch:expr> <bch:> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
1 f1(x2)       32ms 32.5ms      30.6    38.1MB     5.09    12     2      393ms
2 f2(x2)     31.6ms 32.4ms      30.9    38.1MB     8.42    11     3      356ms
3 f3(x2)     33.5ms   35ms      28.5    38.1MB     2.20    13     1      455ms
# … with 4 more variables: result <list>, memory <list>, time <list>,
#   gc <list>
```

Reply:
It's interesting that if I run the code interactively, `f3()` is always the fastest. This is what I expected because `storage.mode<-` calls a primitive R function. Moreover, since the only difference is about the attributes, it shouldn't have such huge (2x) impact on the performance....

But if I use reprex to run, it's always the slowest. Here's what I get if run interactively.

<img width="852" alt="image" src="https://user-images.githubusercontent.com/8368933/60757583-2ffcef00-a03f-11e9-8503-c0fb6e3c1d5f.png">

Reply:
I believe you are right. After some experiments, I realize that use your version (check attributes if NULL and then copy the attributes) is the best choice. `storage.mode<-` is much slower if the vector is lengthy (I have no idea why it's the case). Moreover, use `storage.mode<-` is kind of confusing.